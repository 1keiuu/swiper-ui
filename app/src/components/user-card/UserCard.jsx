import React, { useEffect, useState } from "react";
import "./UserCard.scss"
const UserCard = (props) => {
    const user = props.user
  return (
        <div key={user.id} className="user-card">
          <img
            src={user.imageURL}
            width={260}
            height={260}
            className="user-card__img"
          />
          <div className="user-card__info">
            <p>{user.name},</p>
            <p>{user.age}</p>
          </div>
        </div>
  );
};

export default UserCard;
