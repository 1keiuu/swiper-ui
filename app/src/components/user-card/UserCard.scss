$topPadding: 50px;

.user-card {
  background: white;
  position: absolute;
  height: 400px;
  display: none;
  width: 90%;
  margin: auto;
  top: 0;
  bottom: 0;
  &.--current {
    z-index: 5;
    display: block;
  }
  &.--prev {
    z-index: 6;
    display: block;
  }
  &.--like {
    animation: like 1s ease-out forwards;
  }
  &.--nope {
    animation: nope 1s ease-out forwards;
  }
  .user-card__inner {
    position: relative;
    width: 100%;
    height: 100%;
    -webkit-perspective: 1000px;
    perspective: 1000px;
    display: flex;
    flex-direction: column;
    align-items: center;
    &.--flipped {
      .user-card__item {
        &.user-card__front {
          transform: rotateY(180deg);
        }
        &.user-card__back {
          transform: rotateY(0);
        }
      }
    }
  }

  .user-card__item {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 400px;
    cursor: pointer;
    transition: transform 1s;
    -webkit-backface-visibility: hidden;
    backface-visibility: hidden;
    background: white;
    box-shadow: 0 1px 5px rgb(0 0 0 / 20%);
    border-radius: 8px;

    &.user-card__front {
      .user-card__img {
        object-fit: cover;
        border-top-left-radius: 8px;
        border-top-right-radius: 8px;
        width: 100%;
      }
      .user-card__info {
        display: flex;
        background: white;
        width: 100%;
        border-radius: 8px;
        .name-text {
          padding-left: 10px;
          font-weight: bold;
          margin-right: 8px;
        }
        .age-text {
          font-weight: bold;
        }
      }
    }
    &.user-card__back {
      border-radius: 8px;
      transform: rotateY(-180deg);
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
}

@keyframes like {
  0% {
    right: 0;
    transform: scale(1.05) rotate(0deg);
    display: block;
  }
  100% {
    right: -200%;
    transform: scale(1.05) rotate(15deg);
    display: none;
  }
}

@keyframes nope {
  0% {
    left: 0;
    transform: scale(1.05) rotate(0deg);
    display: block;
  }
  100% {
    left: -200%;
    transform: scale(1.05) rotate(-15deg);
    display: none;
  }
}
