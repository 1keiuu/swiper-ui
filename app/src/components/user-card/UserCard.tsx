import React, { useContext, useEffect, useState } from "react";
import "./UserCard.scss";
import {
  useCurrentUserCardStateContext,
  useCurrentUserCardDispatchContext,
} from "../../context/CurrentUserCardContext";

type UserCardProps = {
  user: UserCard;
  currentCardIndex: number;
  index: number;
};

const UserCard: React.FC<UserCardProps> = (props) => {
  const user = props.user;
  const currentCardIndex = props.currentCardIndex;
  const index = props.index;

  const currentUserCardState = useCurrentUserCardStateContext();
  const currentUserCardDispatcher = useCurrentUserCardDispatchContext();

  let activeClassText: string = "";

  // NOTE: スワイプ判定用
  let moveX;
  let startX;
  // NOTE: 最低スワイプ量
  const dist = 70;

  const userCardClass = (text: string): string[] => {
    // NOTE: カードの状態(current / next / prev /like / nope)の判定をしてclassを適用する
    const isCurrent = currentCardIndex == index;
    const isPrev = currentCardIndex - 1 == index;

    text += "user-card ";
    if (isCurrent) activeClassText = text += "--current ";
    else if (currentCardIndex + 1 == index) activeClassText = text += "--next ";
    else if (isPrev) {
      activeClassText = text += "--prev ";
      if (currentUserCardState.status == "like")
        activeClassText = text += "--like ";
      if (currentUserCardState.status == "nope")
        activeClassText = text += "--nope ";
    }
    return text.split(" ");
  };

  const userCardInnerClass = () => {
    // NOTE: flipしている状態かどうか
    if (currentUserCardState.isFlipped) return "--flipped";
    else return "";
  };

  return (
    <div
      key={user.id}
      className={userCardClass(activeClassText).join(" ")}
      onClick={() => {
        currentUserCardDispatcher.toggleIsFlipped();
      }}
      onTouchStart={(e) => {
        startX = e.touches[0].pageX;
      }}
      onTouchMove={(e) => {
        moveX = e.changedTouches[0].pageX;
      }}
      onTouchEnd={(e) => {
        const isLeftSwipe = startX > moveX && startX > moveX + dist;
        const isRightSwipe = startX < moveX && startX + dist < moveX;

        if (isLeftSwipe) {
          // FIXME: like/nope後の状態管理系処理を共通化したい
          currentUserCardDispatcher.incrementIndex();
          currentUserCardDispatcher.changeStatus("nope");
          currentUserCardDispatcher.setIsFlipped(false);
        } else if (isRightSwipe) {
          currentUserCardDispatcher.incrementIndex();
          currentUserCardDispatcher.changeStatus("like");
          currentUserCardDispatcher.setIsFlipped(false);
        }
        moveX;
        startX;
      }}
    >
      <div className={["user-card__inner", userCardInnerClass()].join(" ")}>
        <div className="user-card__item user-card__front">
          <div className="user-card__status --like">
            <p>LIKE</p>
          </div>
          <div className="user-card__status --nope">
            <p>NOPE</p>
          </div>
          <img
            src={user.imgURL}
            width={330}
            height={330}
            className="user-card__img"
          />
          <div className="user-card__info">
            <p className="name-text">{user.name},</p>
            <p className="age-text">{user.age}</p>
          </div>
        </div>
        <div className="user-card__item user-card__back">
          <p>{user.profile}</p>
        </div>
      </div>
    </div>
  );
};

export default UserCard;
