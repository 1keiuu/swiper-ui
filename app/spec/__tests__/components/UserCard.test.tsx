import React from "react";
import { cleanup, render } from "@testing-library/react";

import UserCard from "../../../src/components/user-card/UserCard";

describe("UserCard Component", () => {
  const valideUser = {
    id: 1,
    name: "test1",
    imgURL: "http://localhost/image",
    age: 21,
    profile: "test1's profile",
  };
  beforeEach(() => {
    cleanup();
  });
  test("is rendered as before.", () => {
    const { asFragment } = render(
      <UserCard user={valideUser} index={1} currentCardIndex={1}></UserCard>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("render front UI correctly.", () => {
    const { asFragment } = render(
      <UserCard user={valideUser} index={1} currentCardIndex={1}></UserCard>
    );
    // HACK: 壊れやすそうなのでいい方法考える
    const node = asFragment();
    const inner = node.firstChild.childNodes[0];
    const userCardItemFront = inner.childNodes[0];
    const userCardStatusBudgeLike = userCardItemFront.childNodes[0];
    const userCardStatusBudgeNope = userCardItemFront.childNodes[1];
    expect(userCardStatusBudgeLike.firstChild.textContent).toBe("LIKE");
    expect(userCardStatusBudgeNope.firstChild.textContent).toBe("NOPE");
    const userCardImage = userCardItemFront.childNodes[2];
    const userCardInfo = userCardItemFront.childNodes[3];
    const userCardName = userCardInfo.childNodes[0];
    const userCarAge = userCardInfo.childNodes[1];
    expect((userCardImage as HTMLImageElement).src).toBe(valideUser.imgURL);
    expect(userCardName.textContent).toBe(`${valideUser.name},`);
    expect(userCarAge.textContent).toBe(String(valideUser.age));
  });
  test("render back UI correctly.", () => {
    const { asFragment } = render(
      <UserCard user={valideUser} index={1} currentCardIndex={1}></UserCard>
    );
    const node = asFragment();
    const inner = node.firstChild.childNodes[0];
    const userCardItemBack = inner.childNodes[1];
    expect(userCardItemBack.firstChild.textContent).toBe(valideUser.profile);
  });
  test("is rendered and has '--prev' class.", () => {
    const prev = render(
      <UserCard user={valideUser} index={1} currentCardIndex={2}></UserCard>
    );
    expect(
      (prev.container.firstChild as HTMLDivElement).classList.contains("--prev")
    );
  });
  test("is rendered and has '--next' class.", () => {
    const prev = render(
      <UserCard user={valideUser} index={1} currentCardIndex={0}></UserCard>
    );
    expect(
      (prev.container.firstChild as HTMLDivElement).classList.contains("--next")
    );
  });
});
