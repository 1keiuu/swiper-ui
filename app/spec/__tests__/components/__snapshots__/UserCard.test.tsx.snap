// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard Component is rendered as before. 1`] = `
<DocumentFragment>
  <div
    class="user-card --current "
  >
    <div
      class="user-card__inner "
    >
      <div
        class="user-card__item user-card__front"
      >
        <div
          class="user-card__status-budge --like"
        >
          <p>
            LIKE
          </p>
        </div>
        <div
          class="user-card__status-budge --nope"
        >
          <p>
            NOPE
          </p>
        </div>
        <img
          alt="user image"
          class="user-card__img"
          height="330"
          src="http://localhost/image"
          width="330"
        />
        <div
          class="user-card__info"
        >
          <p
            class="name-text"
          >
            test1,
          </p>
          <p
            class="age-text"
          >
            21
          </p>
        </div>
      </div>
      <div
        class="user-card__item user-card__back"
      >
        <p>
          test1's profile
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
